source/common/network/utility.cc

```diff
@@ -343,19 +344,32 @@ Address::InstanceConstSharedPtr Utility::getOriginalDst(Socket& sock) {
     return nullptr;
   }
 
-  sockaddr_storage orig_addr;
-  memset(&orig_addr, 0, sizeof(orig_addr));
-  socklen_t addr_len = sizeof(sockaddr_storage);
-  int status;
-
+  SocketOptionName opt_dst;
+  SocketOptionName opt_tp;
   if (*ipVersion == Address::IpVersion::v4) {
-    status = sock.getSocketOption(SOL_IP, SO_ORIGINAL_DST, &orig_addr, &addr_len).return_value_;
+    opt_dst = ENVOY_SOCKET_SO_ORIGINAL_DST;
+    opt_tp = ENVOY_SOCKET_IP_TRANSPARENT;
   } else {
-    status =
-        sock.getSocketOption(SOL_IPV6, IP6T_SO_ORIGINAL_DST, &orig_addr, &addr_len).return_value_;
+    opt_dst = ENVOY_SOCKET_IP6T_SO_ORIGINAL_DST;
+    opt_tp = ENVOY_SOCKET_IPV6_TRANSPARENT;
   }
 
+  sockaddr_storage orig_addr;
+  memset(&orig_addr, 0, sizeof(orig_addr));
+  socklen_t addr_len = sizeof(sockaddr_storage);
+  int status =
+      sock.getSocketOption(opt_dst.level(), opt_dst.option(), &orig_addr, &addr_len).return_value_;
+
   if (status != 0) {
+    if (Api::OsSysCallsSingleton::get().supportsIpTransparent()) {
```

Comment:
Does this make a syscall? Would it be better to save the result once per filter instantiation, instead of calling it per connection?

Reply:
I think the socket syscall would be invoked only once since it's a singleton instance and the return is a static const.

https://github.com/envoyproxy/envoy/blob/5183dbf7f11b23f1f7013a3e42cbce71eb63252a/source/common/api/posix/os_sys_calls_impl.cc#L141